caip-browser_extension_provider_communication.md

```diff
@@ -0,0 +1,141 @@
+  * NOTE: Should mention discovery somewhere since that will require injection still
+* The website author can control when the provider is initialized, and fine-tune performance.
+* No code needs to be injected, so websites would no longer be broken by injected code.
+* The provider library can be published with breaking changes, allowing changes to the API without needing to embed legacy code in each website.
+
+A provider library can be similarly easy to use for website authors as well, requiring nothing more than a single script tag to import a library and get an equivalent experience to using an the injected provider.
+
+Web extension inter-operability is a challenge for embeded providers though. That is what this proposal means to address. Today there is no way to write a provider such that it is compatible with any web extension. Web extensions differ today in how they communicate with wallets, from the messaging system used to the messaging format. These details often aren't publicly documented or treated as a public-facing API, so they can change without notice, making it risky even to embed support for popular conventions used today.
+
+A standard method for providers to communicate with web extebsions would allow website authors to embed their own providers without losing web extension inter-operability.
+
+## Specification
+<!--The technical specification should describe the standard in detail. The specification should be detailed enough to allow competing, interoperable implementations. -->
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
+uppercase in this document are to be interpreted as described in [RFC
+2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Summary
+
+Web extensions should expose a standard interface over [`externally_connectable`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable) to enable the inter-operability of embedded providers.
+
+### Message format
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "/caip-x/schemas/provider-request.schema.json",
+  "title": "Provider / Web Extension message",
+  "description": "A request sent between a provider and a web extension.",
+  "type": "object",
+  "properties": {
+    "type": {
+      "description": "The message type, used to identify this as a CAIP-X message",
+      "const": "caip-x"
+    },
+    "data": {
+      "description": "A JSON-RPC message",
+      "type": "object"
+    }
+  },
+  "required": ["type", "data"]
+}
+```
+
+TODO: should type be more defined here? Seems weird to specify the envelop but no types it can be used with?
+Does `data` need to be a JSON-RPC message specifically?
+Should `type` be something besides `caip-x`?
```

Comment:
Something missing from this CAIP is a specification for what the type for an EVM provider should look like, and the shape of its data

Reply:
not sure if data needs to be an object

Reply:
> I think we need yet another CAIP that specifies what the shape of evm provider data that goes through this envelope

I'm not entirely sure this is necessary

Reply:
actually hmm... so currently we envelope our port stream connection messages, but that is a MetaMask specific implementation detail. I'm not sure that it is something we would want to include in a standard. Perhaps this was Mark's intention and I am only just now understanding/realizing that. This would mean any MetaMask specific messages/streams are still necessary to be flattened and contained within JSON-RPC formatted messages instead. I imagine this won't be an issue, but haven't fully investigated yet

Reply:
Where are we on this thread now?

Reply:
need to update this section to specify what an ethereum provider message should look like for type `caip-x`. `caip-x` should be the caip number that this proposal gets assigned. Future CAIPs can extend this

Reply:
I believe the idea was that the `type` field would specify the `data` format

So for Ethereum it maybe `caip-100` with JSON-RPC in the data field. For Cardono it maybe `caip-101` with GraphQL query in the data field